pow/src/compute/mod.rs

```diff
@@ -89,6 +91,51 @@ fn need_new_vm<M: randomx::WithCacheMode>(
 	need_new_vm
 }
 
+fn recover_allocation_error<M: randomx::WithCacheMode>(
+	key_hash: &H256,
+	machine: &RefCell<Option<(H256, randomx::VM<M>)>>,
+	shared_caches: &mut MutexGuard<LruCache<H256, Arc<randomx::Cache<M>>>>,
+) -> () {	
+	let old_key_hash = machine
+						.borrow()
+						.as_ref()
+						.expect("Any cache has been allocated succesfully.")
+						.0;
+
+	machine.replace(None);
+
+	let last_thread = !(Arc::strong_count((*shared_caches)
+						.iter()
+						.find(|&(key, _)| *key == old_key_hash)
+						.expect("Old key_hash cache should still be in memory.")
+						.1) > 1);
+
+	if last_thread {
+		let mut cache = shared_caches
+			.remove(&old_key_hash)
+			.expect("Old key_hash cache should still be in memory.");
+
+		// Update the cache.
+		Arc::get_mut(&mut cache)
+			.expect("There should be only one reference.")
+			.init(&key_hash[..]);
+
+			shared_caches.insert(*key_hash, cache.clone());
+	} else {
+		while !shared_caches.contains_key(&key_hash) {
+			RECOVERED.wait(shared_caches);
```

Comment:
I really don't like the fact that this can cause a thread to halt (and potentially wait indefinitely). I'd suggest we just return proper error when we cannot do anything and let the thread exits.

Currently halting threads are fine, as they're all mining threads, but this opens a whole box of potential misuses if someone accidentally used `ComputeMode::Mining` on a block import thread.